Add polyfactory framework

adrianeboyd opened the pull request:
Add a framework that generates mock responses using `polyfactory`.

Related to #1.

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

This pull request adds a new framework, PolyfactoryFramework, which generates mock responses using the polyfactory library. Configuration for this framework has been added to config.yaml, and the framework is imported in frameworks/__init__.py.

- **New Features**:
    - Introduced PolyfactoryFramework to generate mock responses using the polyfactory library.
- **Enhancements**:
    - Updated config.yaml to include configuration for PolyfactoryFramework.
    - Modified frameworks/__init__.py to import PolyfactoryFramework.

<!-- Generated by sourcery-ai[bot]: end summary -->

adrianeboyd committed:
Add polyfactory framework

Add a framework that generates mock responses using `polyfactory`.

adrianeboyd commented:
(This sourcery thing seems noisier and less capable than a linter.)

stephenleo commented:
Hey I ran your branch locally and have some comments:
1. The framework will better suit the `Synthetic Data Generation` task instead of the `Multi-label classification` task. I'm planning it out and will submit some updates to your branch once that's ready.
2. It seems that the framework generates only synthetic `labels` without corresponding `text`. I'm not sure if polyfactory can do that. So we may need to change the response_model used to something more suitable for synthetic data generation. I'm still thinking about this as part of the `Synthetic Data Generation` task too.

So for now, I'll keep this PR open and revisit it again once I have `Synthetic Data Generation` task up and running.

adrianeboyd commented:
I kind of disagree, because I think it's more reasonable to have it available as a comparison for classification tasks that have a sensible random baseline. I think it would be much less interesting for a synthetic generation task (I guess unless the task is boring enough that you should be using faker instead).

It doesn't refer to the input text because it's just generating a random list of labels from the provided schema. (In a few tests where I limited the number of possible labels to match the sampling setup and then sampled more data, it had something like 0.1-0.3% accuracy. I would also guess that a majority baseline might be better than a random baseline, but I didn't try that.)

And this was all a bit facetious, I didn't necessarily expect it to be merged, since adding any accuracy metrics to the table would make you immediately want to eliminate it. The point was just that you can easily generate the structure and a random baseline from the response model.

stephenleo commented:
Ah got it. good point! Yep, I'm definitely on the lookout for a suitable dataset to include an accuracy metric that will immediately flag a random label generator as inaccurate. I'm currently prioritizing getting the code up first so that datasets can be easily swapped in and out.

adrianeboyd committed:
Merge remote-tracking branch 'upstream/main' into add-polyfactory

adrianeboyd committed:
Update config and framework

adrianeboyd committed:
Modify synthetic data postal code field definition

Use an equivalent six digit postal code field definition that is
supported by polyfactory rather than a separate validator method.

adrianeboyd committed:
Add dummy placeholder value to otherwise empty init_kwargs settings

adrianeboyd committed:
Fix requirements

adrianeboyd committed:
Merge branch 'main' into add-polyfactory
